components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,395 @@
+    discover_eth_allowance_callbacks_.push_back(std::move(callback));
+    return;
+  }
+  discover_eth_allowance_callbacks_.push_back(std::move(callback));
+
+  const auto keyring_info =
+      keyring_service_->GetKeyringInfoSync(mojom::kDefaultKeyringId);
+  std::vector<std::string> account_addresses;
+  for (const auto& account_info : keyring_info->account_infos) {
+    account_addresses.push_back(account_info->address);
+  }
+
+  if (account_addresses.empty()) {
+    OnDiscoverEthAllowancesCompleted({});
+    return;
+  }
+
+  allowance_discovery_tasks_.clear();
+  get_block_tasks_ = 0;
+
+  const auto token_list_map =
+      BlockchainRegistry::GetInstance()->GetEthTokenListMap(
+          GetChainIdsForAlowanceDiscovering());
+
+  for (const auto& [chain_id, token_list] : token_list_map) {
+    base::Value::List contract_addresses;
+    for (const auto& token : token_list) {
+      contract_addresses.Append(token->contract_address);
+    }
+
+    if (contract_addresses.empty()) {
+      continue;
+    }
+    auto internal_callback = base::BindOnce(
+        &EthAllowanceManager::OnGetCurrentBlock, weak_ptr_factory_.GetWeakPtr(),
+        chain_id, std::move(contract_addresses), account_addresses);
+    get_block_tasks_++;
+    json_rpc_service_->GetBlockNumber(chain_id, std::move(internal_callback));
+  }
+}
+
+void EthAllowanceManager::OnGetCurrentBlock(
+    const std::string& chain_id,
+    base::Value::List contract_addresses,
+    const std::vector<std::string>& account_addresses,
+    uint256_t block_num,
+    mojom::ProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::ProviderError::kSuccess) {
+    get_block_tasks_--;
+    MaybeMergeAllResultsAndCallBack();
+    return;
+  }
+
+  const auto& allowance_cache_dict =
+      prefs_->GetDict(kBraveWalletEthAllowancesCache);
+
+  const auto approval_topic_hash = KeccakHash(kApprovalTopicFunctionSignature);
+  for (const auto& account_address : account_addresses) {
+    std::string account_address_hex;
+    if (!PadHexEncodedParameter(account_address, &account_address_hex)) {
+      continue;
+    }
+    const int task_id(allowance_discovery_tasks_.size());
+    auto internal_callback =
+        base::BindOnce(&EthAllowanceManager::OnGetAllowances,
+                       weak_ptr_factory_.GetWeakPtr(), task_id);
+    allowance_discovery_tasks_.insert_or_assign(
+        task_id, std::make_unique<EthAllowanceTask>(
+                     task_id, chain_id, account_address_hex, block_num));
+
+    base::Value::List topics;
+    topics.Append(approval_topic_hash);
+    topics.Append(account_address_hex);
+
+    base::Value::Dict filter_options;
+    filter_options.Set(kAddress, contract_addresses.Clone());
+    filter_options.Set(kTopics, std::move(topics));
+    filter_options.Set(
+        kFromBlock, GetBlocknumberFilterFromCache(
+                        allowance_cache_dict, chain_id, account_address_hex));
```

Comment:
GetBlocknumberFilterFromCache should get prefs_ from args (instead of allowance_cache_dict)

Reply:
fixed